clients/drcachesim/tools/internal_record_view.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+std::string
+internal_record_view_t::initialize_shard_type(shard_type_t shard_type)
+{
+    return "";
+}
+
+std::string
+internal_record_view_t::initialize_stream(memtrace_stream_t *serial_stream)
+{
+    /* Print warning header.
+     */
+    std::cerr << "The internal_record_view tool is a pretty printer for trace_entry_t "
+                 "records.\nThe trace_entry_t internal record format is not a public "
+                 "interface and is subject to change without notice.\nThis tool is meant "
+                 "for debugging only, please do not rely on the details here.\n";
+    return "";
+}
+
+void *
+internal_record_view_t::parallel_shard_init_stream(int shard_index, void *worker_data,
+                                                   memtrace_stream_t *shard_stream)
+{
+    return shard_stream;
+}
+
+bool
+internal_record_view_t::parallel_shard_exit(void *shard_data)
+{
+    return true;
+}
+
+std::string
+internal_record_view_t::parallel_shard_error(void *shard_data)
+{
+    return "";
+}
+
+bool
+internal_record_view_t::parallel_shard_memref(void *shard_data,
+                                              const trace_entry_t &entry)
+{
+    if (should_skip())
+        return true;
+
+    trace_type_t trace_type = (trace_type_t)entry.type;
+    if (trace_type == TRACE_TYPE_INVALID) {
+        std::cerr << "ERROR: trace_entry_t invalid.\n";
+        return false;
+    }
+
+    std::string trace_type_name = std::string(trace_type_names[trace_type]);
+
+    /* Large if-else for all TRACE_TYPE_.  Prints one line per trace_entry_t.
+     * In some cases we use some helper functions (e.g., type_is_instr()), which group
+     * similar TRACE_TYPE_ together, otherwise we compare trace_type against one or two
+     * specific TRACE_TYPE_ directly.
+     */
+    if (trace_type == TRACE_TYPE_HEADER) {
+        trace_version_t trace_version = (trace_version_t)entry.addr;
+        std::string trace_version_name = std::string(trace_version_names[trace_version]);
+        std::cerr << "<" << trace_type_name << ", trace_version: " << trace_version
+                  << " == " << trace_version_name << ">\n";
+    } else if (trace_type == TRACE_TYPE_FOOTER) {
+        std::cerr << "<" << trace_type_name << ">\n";
+    } else if ((trace_type == TRACE_TYPE_THREAD) ||
+               (trace_type == TRACE_TYPE_THREAD_EXIT)) {
+        uint tid = (uint)entry.addr;
+        std::cerr << "<" << trace_type_name << ", tid: " << tid << ">\n";
+    } else if (trace_type == TRACE_TYPE_PID) {
+        uint pid = (uint)entry.addr;
+        std::cerr << "<" << trace_type_name << ", pid: " << pid << ">\n";
+    } else if (trace_type == TRACE_TYPE_ENCODING) {
+        unsigned short num_encoding_bytes = entry.size;
+        std::cerr << "<" << trace_type_name
+                  << ", num_encoding_bytes: " << num_encoding_bytes
+                  << ", encoding_bytes: 0x" << std::hex;
+        /* Print encoding byte by byte (little-endian).
```

Comment:
Byte-by-byte is only desirable for x86; we'd want words for AArch64, words or half-words (presumably we have LSB 1 for Thumb) for AArch32, and words for RISCV.